File: php/tests/ClientTest.php
<?php

declare(strict_types=1);

namespace Clarian\Tests;

use Clarian\ClarianException;
use Clarian\Client;
use PHPUnit\Framework\TestCase;

final class ClientTest extends TestCase
{
    public function testEnvDetectionTestPrefix(): void
    {
        $c = new Client('cl_test_sk_abc', 'ws');
        $this->assertSame(Client::BASE_URL_TEST, $c->baseUrl);
    }

    public function testEnvDetectionLivePrefix(): void
    {
        $c = new Client('cl_live_sk_abc', 'ws');
        $this->assertSame(Client::BASE_URL_LIVE, $c->baseUrl);
    }

    public function testEnvDetectionUnknownDefaultsLive(): void
    {
        $c = new Client('other_key', 'ws');
        $this->assertSame(Client::BASE_URL_LIVE, $c->baseUrl);
    }

    public function testBaseUrlOverrideStripsTrailingSlash(): void
    {
        $c = new Client('cl_test_sk_x', 'ws', 'http://localhost:9999/v1/');
        $this->assertSame('http://localhost:9999/v1', $c->baseUrl);
    }

    public function testAuthHeadersOnEveryCall(): void
    {
        $mt = new FakeTransport(static fn () => [200, [
            'ok' => true,
            'environment' => 'sandbox',
            'master_account_id' => 'm1',
        ]]);
        $c = new Client('cl_test_sk_key', 'ws-uuid', 'https://mock', null, $mt);
        $res = $c->ping();
        $this->assertTrue($res['ok']);
        $this->assertSame('GET', $mt->last()->method);
        $this->assertSame('/ping', $mt->last()->path);
        $this->assertSame('Bearer cl_test_sk_key', $mt->last()->headers['Authorization']);
        $this->assertSame('ws-uuid', $mt->last()->headers['X-Workspace-Id']);
    }

    public function testCashInCreateSendsIdempotencyAndWorkspace(): void
    {
        $mt = new FakeTransport(static fn () => [201, [
            'ok' => true,
            'environment' => 'sandbox',
            'order' => [
                'id' => 'ord_1',
                'status' => 'pending',
                'amount' => 19.50,
                'currency' => 'BRL',
                'pix' => ['copy_paste' => 'brcode'],
            ],
        ]]);
        $c = new Client('cl_test_sk_x', 'ws-uuid', 'https://mock', null, $mt);
        $res = $c->cashIn->create([
            'amount' => 19.50,
            'payer' => ['name' => 'Ana', 'document_number' => '52998224725'],
        ], 'pay-ext-1');
        $this->assertSame('ord_1', $res['order']['id']);
        $this->assertSame('POST', $mt->last()->method);
        $this->assertSame('/cash-in/pix', $mt->last()->path);
        $this->assertSame('pay-ext-1', $mt->last()->headers['Idempotency-Key']);
        $this->assertSame('ws-uuid', $mt->last()->headers['X-Workspace-Id']);
        $sent = json_decode((string) $mt->last()->body, true);
        $this->assertSame(19.50, $sent['amount']);
        $this->assertSame('Ana', $sent['payer']['name']);
    }

    public function testCashOutCreateSendsIdempotency(): void
    {
        $mt = new FakeTransport(static fn () => [201, [
            'ok' => true,
            'order' => ['id' => 'out_9', 'status' => 'processing'],
        ]]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $c->cashOut->create([
            'amount' => 10.00,
            'pix_key' => 'abc',
            'pix_key_type' => 'EVP',
        ], 'repasse-1');
        $this->assertSame('repasse-1', $mt->last()->headers['Idempotency-Key']);
        $this->assertSame('/cash-out/pix', $mt->last()->path);
    }

    public function testGetDoesNotSendIdempotencyKey(): void
    {
        $mt = new FakeTransport(static fn () => [200, [
            'ok' => true,
            'order' => ['id' => 'x', 'status' => 'pending'],
        ]]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $c->cashIn->retrieve('x');
        $this->assertArrayNotHasKey('Idempotency-Key', $mt->last()->headers);
    }

    public function testCashInRetrieveHappyPath(): void
    {
        $mt = new FakeTransport(static fn () => [200, [
            'ok' => true,
            'order' => ['id' => 'ord_paid', 'status' => 'completed'],
        ]]);
        $c = new Client('cl_live_sk_x', 'ws', 'https://mock', null, $mt);
        $res = $c->cashIn->retrieve('ord_paid');
        $this->assertSame('completed', $res['order']['status']);
        $this->assertSame('/cash-in/ord_paid', $mt->last()->path);
    }

    public function testCashOutRetrieveHappyPath(): void
    {
        $mt = new FakeTransport(static fn () => [200, [
            'ok' => true,
            'order' => ['id' => 'out_1', 'status' => 'completed'],
        ]]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $res = $c->cashOut->retrieve('out_1');
        $this->assertSame('out_1', $res['order']['id']);
        $this->assertSame('/cash-out/out_1', $mt->last()->path);
    }

    public function testCashOutDictCheckHappyPath(): void
    {
        $mt = new FakeTransport(static fn () => [200, [
            'ok' => true,
            'dict' => ['name' => 'Maria', 'keyType' => 'EMAIL'],
        ]]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $res = $c->cashOut->dictCheck('maria@example.com', 'EMAIL');
        $this->assertSame('Maria', $res['dict']['name']);
        $this->assertSame('POST', $mt->last()->method);
        $this->assertSame('/pix/payouts/dict', $mt->last()->path);
        $sent = json_decode((string) $mt->last()->body, true);
        $this->assertSame('maria@example.com', $sent['pix_key']);
        $this->assertSame('EMAIL', $sent['key_type']);
    }

    public function testBalancesListHappyPath(): void
    {
        $mt = new FakeTransport(static fn () => [200, [
            'ok' => true,
            'balances' => [[
                'currency' => 'BRL',
                'available' => 100.50,
                'pending' => 0,
                'locked' => 0,
            ]],
        ]]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $bals = $c->balances->list();
        $this->assertCount(1, $bals);
        $this->assertSame('BRL', $bals[0]['currency']);
        $this->assertSame('/account/balances', $mt->last()->path);
    }

    public function testTransactionsListAndRetrieveHappyPath(): void
    {
        $mt = new FakeTransport(function (RecordedRequest $req) {
            if (str_starts_with($req->path, '/transactions?') || $req->path === '/transactions') {
                $this->assertStringContainsString('type=pix_in', $req->path);
                $this->assertStringContainsString('limit=20', $req->path);

                return [200, [
                    'transactions' => [[
                        'id' => 'tx1',
                        'type' => 'pix_in',
                        'status' => 'completed',
                        'amount' => 10,
                        'fee' => 0,
                        'currency' => 'BRL',
                    ]],
                ]];
            }
            if ($req->path === '/transactions/tx-99') {
                return [200, [
                    'transaction' => [
                        'id' => 'tx-99',
                        'type' => 'pix_in',
                        'status' => 'completed',
                        'amount' => 10,
                        'fee' => 0.1,
                        'currency' => 'BRL',
                    ],
                ]];
            }
            $this->fail('unexpected path ' . $req->path);
        });
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $lst = $c->transactions->list('pix_in', null, 20);
        $this->assertSame('tx1', $lst[0]['id']);
        $tx = $c->transactions->retrieve('tx-99');
        $this->assertSame('tx-99', $tx['id']);
    }

    public function testWalletsListAndBalanceHappyPath(): void
    {
        $mt = new FakeTransport(function (RecordedRequest $req) {
            if ($req->path === '/wallets' || str_starts_with($req->path, '/wallets?')) {
                return [200, [
                    'ok' => true,
                    'wallets' => [[
                        'id' => 'w1',
                        'network' => 'polygon',
                        'address' => '0xabc',
                    ]],
                ]];
            }
            if ($req->path === '/wallets/w1/balance') {
                return [200, [
                    'wallet_id' => 'w1',
                    'network' => 'polygon',
                    'address' => '0xabc',
                    'balances' => [['currency' => 'USDT', 'amount' => '1.5']],
                ]];
            }
            $this->fail('unexpected path ' . $req->path);
        });
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $wallets = $c->wallets->list();
        $this->assertSame('w1', $wallets[0]['id']);
        $bal = $c->wallets->retrieveBalance('w1');
        $this->assertSame('w1', $bal['wallet_id']);
        $this->assertSame('USDT', $bal['balances'][0]['currency']);
    }

    public function testWebhooksCrudHappyPath(): void
    {
        $mt = new FakeTransport(function (RecordedRequest $req) {
            if ($req->method === 'POST' && $req->path === '/webhooks') {
                return [201, [
                    'ok' => true,
                    'subscription' => [
                        'id' => 'wh1',
                        'url' => 'https://ex.com',
                        'events' => ['pix_payin.completed'],
                        'is_active' => true,
                    ],
                    'secret' => 'whsec_abc',
                ]];
            }
            if ($req->method === 'GET' && $req->path === '/webhooks') {
                return [200, [
                    'subscriptions' => [[
                        'id' => 'wh1',
                        'url' => 'https://ex.com',
                        'events' => ['pix_payin.completed'],
                        'is_active' => true,
                    ]],
                ]];
            }
            if ($req->method === 'PATCH' && $req->path === '/webhooks/wh1') {
                return [200, [
                    'subscription' => [
                        'id' => 'wh1',
                        'url' => 'https://ex.com/v2',
                        'events' => ['pix_payin.completed'],
                        'is_active' => false,
                    ],
                ]];
            }
            if ($req->method === 'DELETE' && $req->path === '/webhooks/wh1') {
                return [200, ['ok' => true]];
            }
            $this->fail('unexpected ' . $req->method . ' ' . $req->path);
        });
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        $created = $c->webhooks->create([
            'url' => 'https://ex.com',
            'events' => ['pix_payin.completed'],
        ]);
        $this->assertSame('whsec_abc', $created['secret']);
        $this->assertSame('wh1', $created['id']);
        $subs = $c->webhooks->list();
        $this->assertCount(1, $subs);
        $upd = $c->webhooks->update('wh1', [
            'url' => 'https://ex.com/v2',
            'events' => ['pix_payin.completed'],
            'is_active' => false,
        ]);
        $this->assertFalse($upd['is_active']);
        $c->webhooks->delete('wh1');
        $this->assertSame('DELETE', $mt->calls[array_key_last($mt->calls)]->method);
    }

    public function testErrorJsonCodeAndMeta(): void
    {
        $mt = new FakeTransport(static fn () => [402, [
            'error' => 'insufficient_balance',
            'detail' => 'saldo insuficiente',
            'available' => 10,
            'requested' => 100,
        ]]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        try {
            $c->cashOut->create(['amount' => 100], 'idem');
            $this->fail('expected ClarianException');
        } catch (ClarianException $err) {
            $this->assertSame(402, $err->status);
            $this->assertSame('insufficient_balance', $err->code);
            $this->assertSame('saldo insuficiente', $err->message);
            $this->assertSame(10, $err->meta['available']);
            $this->assertStringContainsString('insufficient_balance', $err->getMessage());
        }
    }

    public function testBodyTruncationInMessage(): void
    {
        $longBody = str_repeat('x', 600);
        $mt = new FakeTransport(static fn () => [429, $longBody]);
        $c = new Client('cl_test_sk_x', 'ws', 'https://mock', null, $mt);
        try {
            $c->ping();
            $this->fail('expected ClarianException');
        } catch (ClarianException $err) {
            $this->assertSame(429, $err->status);
            $this->assertSame('', $err->code);
            $this->assertSame(500, strlen($err->message));
            $this->assertLessThanOrEqual(500 + strlen('HTTP 429: '), strlen($err->getMessage()));
            $this->assertStringNotContainsString(str_repeat('x', 600), $err->getMessage());
            $this->assertSame($longBody, $err->body);
        }
    }
}

File: go/client_test.go
package clarian

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestNew_envInference(t *testing.T) {
	t.Parallel()
	cases := []struct {
		name   string
		apiKey string
		want   string
	}{
		{"test prefix", "cl_test_sk_abc", BaseURLTest},
		{"live prefix", "cl_live_sk_abc", BaseURLLive},
		{"unknown defaults live", "other_key", BaseURLLive},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			c := New(tc.apiKey, "ws")
			if c.baseURL != tc.want {
				t.Fatalf("baseURL = %q, want %q", c.baseURL, tc.want)
			}
		})
	}
}

func TestNew_WithBaseURLOverrides(t *testing.T) {
	t.Parallel()
	c := New("cl_test_sk_x", "ws", WithBaseURL("http://localhost:9999/v1/"))
	if c.baseURL != "http://localhost:9999/v1" {
		t.Fatalf("baseURL = %q", c.baseURL)
	}
}

func TestAuthHeadersOnEveryCall(t *testing.T) {
	t.Parallel()
	var gotAuth, gotWS, gotMethod, gotPath string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		gotWS = r.Header.Get("X-Workspace-Id")
		gotMethod = r.Method
		gotPath = r.URL.Path
		_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","master_account_id":"m1"}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_key", "ws-uuid", WithBaseURL(srv.URL))
	if _, err := c.Ping(context.Background()); err != nil {
		t.Fatalf("Ping: %v", err)
	}
	if gotAuth != "Bearer cl_test_sk_key" {
		t.Errorf("Authorization = %q", gotAuth)
	}
	if gotWS != "ws-uuid" {
		t.Errorf("X-Workspace-Id = %q", gotWS)
	}
	if gotMethod != http.MethodGet || gotPath != "/ping" {
		t.Errorf("request = %s %s", gotMethod, gotPath)
	}
}

func TestCashInCreate_idempotencyAndPayer(t *testing.T) {
	t.Parallel()
	var gotIdem, gotBody string
	var gotAuth, gotWS string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		gotWS = r.Header.Get("X-Workspace-Id")
		gotIdem = r.Header.Get("Idempotency-Key")
		raw, _ := io.ReadAll(r.Body)
		gotBody = string(raw)
		if r.URL.Path != "/cash-in/pix" || r.Method != http.MethodPost {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","order":{"id":"ord_1","status":"pending","amount":19.50,"currency":"BRL","pix":{"qr_code":"qr","copy_paste":"brcode","end_to_end_id":null},"fee":0.50}}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws-uuid", WithBaseURL(srv.URL))
	res, err := c.CashIn.Create(context.Background(), "pay-ext-1", CashInRequest{
		Amount: json.Number("19.50"),
		Payer:  Payer{Name: "Ana", DocumentNumber: "52998224725"},
	})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if res.Order.ID != "ord_1" || res.Order.Status != TxStatusPending {
		t.Fatalf("order = %+v", res.Order)
	}
	if res.Order.Pix.CopyPaste == nil || *res.Order.Pix.CopyPaste != "brcode" {
		t.Fatalf("pix = %+v", res.Order.Pix)
	}
	if gotAuth != "Bearer cl_test_sk_x" || gotWS != "ws-uuid" {
		t.Errorf("auth headers: %q %q", gotAuth, gotWS)
	}
	if gotIdem != "pay-ext-1" {
		t.Errorf("Idempotency-Key = %q", gotIdem)
	}

	var body map[string]json.RawMessage
	if err := json.Unmarshal([]byte(gotBody), &body); err != nil {
		t.Fatalf("body: %v", err)
	}
	if string(body["amount"]) != "19.50" {
		t.Errorf("amount wire = %s, want number 19.50", body["amount"])
	}
	var payer Payer
	if err := json.Unmarshal(body["payer"], &payer); err != nil {
		t.Fatalf("payer: %v", err)
	}
	if payer.Name != "Ana" || payer.DocumentNumber != "52998224725" {
		t.Errorf("payer = %+v", payer)
	}
}

func TestCashOutCreate_idempotencyAndPixKeyType(t *testing.T) {
	t.Parallel()
	var gotIdem, gotBody string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotIdem = r.Header.Get("Idempotency-Key")
		raw, _ := io.ReadAll(r.Body)
		gotBody = string(raw)
		w.WriteHeader(http.StatusCreated)
		_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","order":{"id":"out_9","status":"processing","amount":10.00}}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	out, err := c.CashOut.Create(context.Background(), "repasse-1", CashOutRequest{
		Amount:     json.Number("10.00"),
		PixKey:     "abc-key",
		PixKeyType: PixKeyEVP,
	})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if out.Order.ID != "out_9" || out.Order.Status != TxStatusProcessing {
		t.Fatalf("order = %+v", out.Order)
	}
	if gotIdem != "repasse-1" {
		t.Errorf("Idempotency-Key = %q", gotIdem)
	}
	if !strings.Contains(gotBody, `"pix_key_type":"EVP"`) {
		t.Errorf("body missing EVP: %s", gotBody)
	}
	if !strings.Contains(gotBody, `"amount":10.00`) && !strings.Contains(gotBody, `"amount":10`) {
		t.Errorf("amount not numeric: %s", gotBody)
	}
}

func TestCashInGet(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/cash-in/ord_paid" {
			t.Errorf("path = %s", r.URL.Path)
		}
		_, _ = w.Write([]byte(`{"ok":true,"order":{"id":"ord_paid","status":"completed","amount":19.50}}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_live_sk_x", "ws", WithBaseURL(srv.URL))
	res, err := c.CashIn.Get(context.Background(), "ord_paid")
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if res.Order.Status != TxStatusCompleted {
		t.Fatalf("status = %s", res.Order.Status)
	}
}

func TestBalancesList(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/account/balances" {
			t.Errorf("path = %s", r.URL.Path)
		}
		_, _ = w.Write([]byte(`{"ok":true,"balances":[{"currency":"BRL","available":100.50,"pending":0,"locked":0}]}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	bals, err := c.Balances.List(context.Background())
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(bals) != 1 || bals[0].Currency != "BRL" || bals[0].Available != "100.50" {
		t.Fatalf("balances = %+v", bals)
	}
}

func TestTransactionsListAndGet(t *testing.T) {
	t.Parallel()
	mux := http.NewServeMux()
	mux.HandleFunc("/transactions/", func(w http.ResponseWriter, r *http.Request) {
		id := strings.TrimPrefix(r.URL.Path, "/transactions/")
		_, _ = w.Write([]byte(`{"transaction":{"id":"` + id + `","type":"pix_in","status":"completed","amount":10,"fee":0.1,"currency":"BRL","created_at":"t","updated_at":"t"}}`))
	})
	mux.HandleFunc("/transactions", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Query().Get("type") != "pix_in" || r.URL.Query().Get("limit") != "20" {
			t.Errorf("query = %s", r.URL.RawQuery)
		}
		_, _ = w.Write([]byte(`{"transactions":[{"id":"tx1","type":"pix_in","status":"completed","amount":10,"fee":0,"currency":"BRL","created_at":"t","updated_at":"t"}]}`))
	})
	srv := httptest.NewServer(mux)
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	list, err := c.Transactions.List(context.Background(), ListTransactionsParams{Type: "pix_in", Limit: 20})
	if err != nil {
		t.Fatalf("List: %v", err)
	}
	if len(list) != 1 || list[0].ID != "tx1" {
		t.Fatalf("list = %+v", list)
	}
	tx, err := c.Transactions.Get(context.Background(), "tx-99")
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if tx.ID != "tx-99" {
		t.Fatalf("tx = %+v", tx)
	}
}

func TestWebhooksCRUD(t *testing.T) {
	t.Parallel()
	// Fixtures match live api-gateway wrappers (subscription/secret siblings).
	mux := http.NewServeMux()
	mux.HandleFunc("/webhooks/", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","subscription":{"id":"wh1","url":"https://ex.com","events":["pix_payin.completed"],"is_active":true,"description":null,"environment":"sandbox","created_at":"t","updated_at":"t2"}}`))
		case http.MethodPatch:
			_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","subscription":{"id":"wh1","url":"https://ex.com/v2","events":["pix_payin.completed"],"is_active":false,"description":null,"environment":"sandbox","created_at":"t","updated_at":"t3"}}`))
		case http.MethodDelete:
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"ok":true}`))
		default:
			t.Errorf("method %s", r.Method)
		}
	})
	mux.HandleFunc("/webhooks", func(w http.ResponseWriter, r *http.Request) {
		switch r.Method {
		case http.MethodGet:
			_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","subscriptions":[{"id":"wh1","url":"https://ex.com","events":["pix_payin.completed"],"is_active":true,"created_at":"t","updated_at":"t2"}]}`))
		case http.MethodPost:
			w.WriteHeader(http.StatusCreated)
			_, _ = w.Write([]byte(`{"ok":true,"environment":"sandbox","subscription":{"id":"wh1","url":"https://ex.com","events":["pix_payin.completed"],"is_active":true,"description":null,"environment":"sandbox","created_at":"t","updated_at":"t"},"secret":"whsec_abc"}`))
		default:
			t.Errorf("method %s", r.Method)
		}
	})
	srv := httptest.NewServer(mux)
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	ctx := context.Background()

	created, err := c.Webhooks.Create(ctx, WebhookInput{URL: "https://ex.com", Events: []string{"pix_payin.completed"}})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if created.Secret != "whsec_abc" || created.ID != "wh1" {
		t.Fatalf("created = %+v secret=%q", created.Webhook, created.Secret)
	}

	subs, err := c.Webhooks.List(ctx)
	if err != nil || len(subs) != 1 {
		t.Fatalf("List: %v %+v", err, subs)
	}
	got, err := c.Webhooks.Get(ctx, "wh1")
	if err != nil || got.ID != "wh1" || got.UpdatedAt != "t2" {
		t.Fatalf("Get: %v %+v", err, got)
	}
	active := false
	upd, err := c.Webhooks.Update(ctx, "wh1", WebhookInput{URL: "https://ex.com/v2", Events: []string{"pix_payin.completed"}, IsActive: &active})
	if err != nil || upd.IsActive || upd.URL != "https://ex.com/v2" {
		t.Fatalf("Update: %v %+v", err, upd)
	}
	if err := c.Webhooks.Delete(ctx, "wh1"); err != nil {
		t.Fatalf("Delete: %v", err)
	}
}

func TestAPIError_jsonCode(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusPaymentRequired)
		_, _ = w.Write([]byte(`{"error":"insufficient_balance","detail":"saldo insuficiente"}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	_, err := c.CashOut.Create(context.Background(), "idem", CashOutRequest{Amount: json.Number("100")})
	if err == nil {
		t.Fatal("expected error")
	}
	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("want *APIError, got %T %v", err, err)
	}
	if apiErr.Status != 402 || apiErr.Code != "insufficient_balance" {
		t.Fatalf("apiErr = %+v", apiErr)
	}
	if !strings.Contains(apiErr.Body, "insufficient_balance") {
		t.Errorf("Body = %q", apiErr.Body)
	}
}

func TestAPIError_nonJSONBody(t *testing.T) {
	t.Parallel()
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusTooManyRequests)
		_, _ = w.Write([]byte(`rate limited`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	err := c.do(context.Background(), http.MethodGet, "/ping", "", nil, nil)
	if err == nil {
		t.Fatal("expected error")
	}
	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("want *APIError, got %T %v", err, err)
	}
	if apiErr.Status != 429 || apiErr.Code != "" || apiErr.Body != "rate limited" {
		t.Fatalf("apiErr = %+v", apiErr)
	}
}

func TestDo_doesNotSetIdempotencyOnGET(t *testing.T) {
	t.Parallel()
	var gotIdem string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotIdem = r.Header.Get("Idempotency-Key")
		_, _ = w.Write([]byte(`{"ok":true,"order":{"id":"x","status":"pending"}}`))
	}))
	t.Cleanup(srv.Close)

	c := New("cl_test_sk_x", "ws", WithBaseURL(srv.URL))
	if _, err := c.CashIn.Get(context.Background(), "x"); err != nil {
		t.Fatalf("Get: %v", err)
	}
	if gotIdem != "" {
		t.Errorf("unexpected Idempotency-Key on GET: %q", gotIdem)
	}
}
